### DISCUSSION TITLE
Forced Texture pack

### DISCUSSION THREAD

**Discussion by sqcred** [Minestom/Minestom#783] [Help] [answered]
How do i set the texture pack of a player when they join + force it (kick the player if they deny it)

**Comment by MrGazdag** [comment, marked as answer]
You can create a [`net.minestom.server.resourcepack.ResourcePack`](https://javadoc.minestom.net/net/minestom/server/resourcepack/ResourcePack.html) object with the static methods inside `ResourcePack`:
```java
String url = ...;
String hash = ...; //or null if no hash
Component prompt = ...; //or null if no prompt
ResourcePack pack;
pack = ResourcePack.optional(url, hash);
pack = ResourcePack.optional(url, hash, prompt);
pack = ResourcePack.forced(url, hash);         //this is what you want
pack = ResourcePack.forced(url, hash, prompt); //this is what you want
```
To apply a `ResourcePack` to a `Player`, use [`Player#setResourcePack(ResourcePack)`](https://javadoc.minestom.net/net/minestom/server/entity/Player#setResourcePack(net.minestom.server.resourcepack.ResourcePack))

If you use one of the methods called `forced` to create a `ResourcePack`, the client will automatically disconnect itself if the user declines the resource pack. If the pack cannot be downloaded/applied, the client will not disconnect itself. You can detect this, by listening to the [`net.minestom.server.event.player.PlayerResourcePackStatusEvent`](https://javadoc.minestom.net/net/minestom/server/event/player/PlayerResourcePackStatusEvent.html), and checking its `ResourcePackStatusEvent`.

**Reply by emortaldev** [reply]
Is forced more of a client side disconnect vs server disconnect? I assume you still want the server to disconnect them
Or does it just change the deny message to disconnect or something lol

**Reply by Moulberry** [reply]
Forced will make the client disconnect itself if it clicks "deny" or presses ESC, but it won't disconnect if the client fails to download the pack. If you want to make sure the player has the pack, you should use both `forced` and respond to the PlayerResourcePackStatusEvent
